README.md
# Toffee

[![Travis build status](https://travis-ci.org/seankross/toffee.svg?branch=master)](https://travis-ci.org/seankross/toffee) 
[![CRAN status](https://www.r-pkg.org/badges/version/toffee)](https://cran.r-project.org/package=toffee)

## Installation

```r
devtools::install_github("seankross/toffee")
```

## Usage

```r
##################
## toffee_tbl() ##
##################

library(dplyr)

set.seed(2017-10-15)

trees %>%
  mutate(Branches = sample(15:25, size = 31, replace = TRUE)) %>%
  lm(Volume ~ Girth + Height + Branches, data = .) %>%
  toffee_tbl()

# term               Coef               CI p.value
# (Intercept) -67.1978389 [-85.93, -48.46] < 0.01*
# Girth         4.7531349     [4.26, 5.25] < 0.01*
# Height        0.3311483     [0.09, 0.58]   0.01*
# Branches      0.4600241    [-0.01, 0.93]    0.07

#################
## toffee_ci() ##
#################

toffee_ci(c(1, 2, 3), c(.2, .25, .3), I)
# "[0.61, 1.39]" "[1.51, 2.49]" "[2.41, 3.59]"

toffee_ci(c(4, 2, 7), c(.15, .2, .1), exp)
# "[3.71, 4.29]" "[1.61, 2.39]" "[6.8, 7.2]"
```


NEWS.md
# toffee 0.1.0

- First release.
- `toffee_tbl`: create a tibble


tests/testthat.R
library(testthat)
library(toffee)
library(dplyr)

test_check("toffee")


tests/testthat/test-toffee_tbl.R
context("test-toffee_tbl")

test_that("toffee_tbl throws correct errors", {
  expect_error(toffee_tbl(1))
})

# test_that("toffee_tbl makes appropriate tables", {
#
# })


R/toffee_tbl.R
#' Create a nicely formatted table from a model object.
#'
#' Given a model object (like from [stats::lm()], [stats::glm()], or
#' [lme4::glmer()]) this function makes a table intended for interactive use or
#' publication. The model coefficients and confidence intervals displayed in the
#' table have had their [inverse link function](https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function)
#' applied, therefore they are in the units of the response variable.
#'
#' @param model A model object that is compatible with [broom::tidy()].
#' @param conf_level The confidence level that will be used for computing the
#' confidence interval (for example: `0.95` or `0.99`).
#' @param digits The number of decimal places that numbers in the table should
#' be rounded to. For no rounding use `Inf`. This argument is provided to
#' [base::round()]. The default value is `2`.
#' @param concat_signif If `TRUE` the resulting data frame will append
#' significance symbols as generated by [toffee_signif()]. If `FALSE` the
#' significance symbols will appear in a separate column called `Significant`.
#' @param odds_ratio If `TRUE` and the family of the model provided is
#' `logit` then the model coefficients and confidence intervals returned in
#' the resulting table will be odds ratios.
#' @param ci_fmt If `TRUE` then the confidence interval will be formatted between
#' brackets in one column called `CI`. If `FALSE` the lower and upper bounds of
#' the confidence interval will be in their own columns called `Lower_CI` and
#' `Upper_CI` respectfully.
#' @param ... Arguments that will be passed to [toffee_signif()].
#' @return A [tibble::tibble()] with the following columns: the
#' name of the coefficient in the model (`Variable`), the value of the
#' coefficient with inverse of the link function applied (`Estimate` or
#' `Odds_Ratio`), the confidence interval for the coefficient (`CI`), the
#' p-value for the estimate, (`p_value`), and optionally symbols representing
#' levels of significance (`Significant`).
#'
#' @export
#' @importFrom broom tidy
#' @importFrom purrr map_dbl map_chr is_null map2_chr
#' @importFrom dplyr mutate mutate_at select rename "%>%"
#' @examples
#'
#' library(dplyr)
#' library(toffee)
#'
#' linear_model <- toffee_forest %>%
#'   lm(Volume ~ Girth + Height + Branches, data = .)
#'
#' # Basic usage
#' linear_model %>%
#'   toffee_tbl()
#'
#'  # # A tibble: 4 x 4
#'  #   Variable    Estimate CI               p_value
#'  #   <chr>          <dbl> <chr>            <chr>
#'  # 1 (Intercept)   -67.2  [-86.81, -47.59] < 0.01***
#'  # 2 Girth           4.75 [4.23, 5.27]     < 0.01***
#'  # 3 Height          0.33 [0.08, 0.59]     0.01*
#'  # 4 Branches        0.46 [-0.03, 0.95]    0.07
#'
#' # Using the `chars` argument from toffee_signif
#' linear_model %>%
#'   toffee_tbl(thresholds = 0.01, chars = "*")
#'
#'  # # A tibble: 4 x 4
#'  #   Variable    Estimate CI               p_value
#'  #   <chr>          <dbl> <chr>            <chr>
#'  # 1 (Intercept)   -67.2  [-86.81, -47.59] < 0.01*
#'  # 2 Girth           4.75 [4.23, 5.27]     < 0.01*
#'  # 3 Height          0.33 [0.08, 0.59]     0.01
#'  # 4 Branches        0.46 [-0.03, 0.95]    0.07
#'
#' # Separating the significance symbols into their own column
#' linear_model %>%
#'   toffee_tbl(concat_signif = FALSE)
#'
#'  # # A tibble: 4 x 5
#'  #   Variable    Estimate CI               p_value Significant
#'  #   <chr>          <dbl> <chr>            <chr>   <chr>
#'  # 1 (Intercept)   -67.2  [-86.81, -47.59] < 0.01  ***
#'  # 2 Girth           4.75 [4.23, 5.27]     < 0.01  ***
#'  # 3 Height          0.33 [0.08, 0.59]     0.01    *
#'  # 4 Branches        0.46 [-0.03, 0.95]    0.07    ""
#'
#' logistic_model <- toffee_forest %>%
#'   glm(Healthy ~ Girth + Height + Branches + Volume, data = .,
#'       family = binomial())
#'
#' logistic_model %>%
#'   toffee_tbl()
#'
#'  # # A tibble: 5 x 4
#'  #   Variable    Odds_Ratio CI                  p_value
#'  #   <chr>            <dbl> <chr>               <chr>
#'  # 1 (Intercept)       0.05 [0, 10626998633.06] 0.81
#'  # 2 Girth             0.75 [0.14, 3.65]        0.72
#'  # 3 Height            0.92 [0.73, 1.14]        0.45
#'  # 4 Branches          2    [1.31, 3.82]        0.01**
#'  # 5 Volume            1.03 [0.75, 1.45]        0.86
#'
toffee_tbl <- function(model, conf_level = 0.95, digits = 2,
                       concat_signif = TRUE, odds_ratio = TRUE,
                       ci_fmt = TRUE, ...) {

  tryCatch(family(model),
    error = function(c) {
      stop("This model object is not compatible with toffee_tbl.", call. = FALSE)
    }
  )

  link <- family(model)$link

  if(link == "logit" && odds_ratio) {
    mean_function <- exp
  } else {
    mean_function <- make.link(link)$linkinv
  }

  result <- tidy(model, conf.int = TRUE, conf.level = conf_level) %>%
    mutate(Significant = toffee_signif(p.value, ...)) %>%
    mutate_at(c("estimate", "conf.low", "conf.high"), mean_function) %>%
    mutate_at(c("estimate", "conf.low", "conf.high", "p.value"),
              round, digits = digits) %>%
    mutate(p.value = map_chr(p.value, as.character)) %>%
    mutate(p.value = map_chr(p.value, ~ if_else(.x < 0.01, "< 0.01", .x)))

  if (ci_fmt) {
    result <- result %>%
      mutate(CI = map2_chr(conf.low, conf.high,
                           ~ paste0("[", .x, ", ", .y, "]"))) %>%
      dplyr::select(term, estimate, CI, p.value, Significant) %>%
      rename(Variable = term, Estimate = estimate, p_value = p.value)
  } else {
    result <- result %>%
      dplyr::select(term, estimate, conf.low, conf.high, p.value, Significant) %>%
      rename(Variable = term, Estimate = estimate, Lower_CI = conf.low,
             Upper_CI = conf.high, p_value = p.value)
  }

  if(link == "logit" && odds_ratio) {
    result <- result %>%
      rename(Odds_Ratio = Estimate)
  }

  if(concat_signif) {
    result <- result %>%
      mutate(p_value = paste0(p_value, Significant)) %>%
      select(-Significant)
  }

  result
}


R/toffee_signif.R
#' Specify symbols to represent confidence levels.
#'
#' @param p_values A numeric vector of p-values.
#' @param thresholds Descending p-value thresholds. Values less than or equal to
#'  a threshold will produce the corresponding character.
#' @param chars A vector of strings that symbolize a p-value less than the
#' threshold.
#' @export
#' @examples
#'
#' toffee_signif(c(1, .5, .049, .01, .0012, 0))
toffee_signif <- function(p_values,
                         thresholds = c(0.05, 0.01, 0.001),
                         chars = c("*", "**", "***")) {

  if (length(thresholds) != length(chars)) {
    stop("[toffee_signif] Please ensure that there are the same number of thresholds and character symbols.",
         call. = FALSE)
  }

  if (is.unsorted(rev(thresholds))) {
    stop("[toffee_signif] Please ensure that the thresholds are strictly descending.",
         call. = FALSE)
  }

  if (length(thresholds) != length(unique(thresholds))) {
    stop("[toffee_signif] Please ensure that the thresholds contain no repeated values.",
         call. = FALSE)
  }

  symnum(p_values, cutpoints = c(0, rev(thresholds), 1),
         symbols = c(rev(chars), ""),
         corr = FALSE, na = FALSE) %>%
    as.character()
}


R/toffee_ci.R
#' Make pretty two-sided 95pct confidence intervals
#'
#' @param m A vector of sample means.
#' @param se A vector of sample standard errors.
#' @param cf A compliment to a link function. For example use
#' \code{\link[base]{I}} for linear models or \code{\link[base]{exp}} for
#' a binomial model.
#' @return A vector of strings with formatted confidence intervals.
#' @importFrom dplyr data_frame mutate pull "%>%"
#' @importFrom purrr map_dbl map_chr compose map2_chr
#' @export
#' @examples
#' toffee_ci(c(1, 2, 3), c(.2, .25, .3), I)
#' # "[0.61, 1.39]" "[1.51, 2.49]" "[2.41, 3.59]"
#'
#' toffee_ci(c(4, 2, 7), c(.15, .2, .1), exp)
#' # "[3.71, 4.29]" "[1.61, 2.39]" "[6.8, 7.2]"
toffee_ci <- function(m, se, cf) {
  data_frame(m = m, se = se) %>%
    mutate(lower = m - qnorm(.975) * se) %>%
    mutate(upper = m + qnorm(.975) * se) %>%
    mutate(lower = map_dbl(lower, cf)) %>%
    mutate(upper = map_dbl(upper, cf)) %>%
    mutate(lower = map_dbl(lower, round, digits = 2)) %>%
    mutate(upper = map_dbl(upper, round, digits = 2)) %>%
    mutate(ci = map2_chr(lower, upper, ~ paste0("[", .x, ", ", .y, "]"))) %>%
    pull(ci)
}
